core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -502,6 +503,9 @@ public boolean equals(Object o) {
         if (relocatingNodeId != null ? !relocatingNodeId.equals(that.relocatingNodeId) : that.relocatingNodeId != null) {
             return false;
         }
+        if (allocationId != null ? !allocationId.equals(that.allocationId) : that.allocationId != null) {
+            return false;
+        }
```

Comment:
this change requires going over all the places we use `equals` in ShardRouting....